File: experiments/03_synth_data_generation/py_src/batch03_array_ops.py
"""Batch 03: Array/list operations — pure functions."""

def sum_array(arr: list[int]) -> int:
    s = 0
    for x in arr:
        s += x
    return s

def product_array(arr: list[int]) -> int:
    p = 1
    for x in arr:
        p *= x
    return p

def max_element(arr: list[int]) -> int:
    m = arr[0]
    for i in range(1, len(arr)):
        if arr[i] > m:
            m = arr[i]
    return m

def min_element(arr: list[int]) -> int:
    m = arr[0]
    for i in range(1, len(arr)):
        if arr[i] < m:
            m = arr[i]
    return m

def reverse_array(arr: list[int]) -> list[int]:
    result = arr[:]
    n = len(result)
    for i in range(n // 2):
        result[i], result[n - 1 - i] = result[n - 1 - i], result[i]
    return result

def is_sorted(arr: list[int]) -> bool:
    for i in range(len(arr) - 1):
        if arr[i] > arr[i + 1]:
            return False
    return True

def count_occurrences(arr: list[int], val: int) -> int:
    c = 0
    for x in arr:
        if x == val:
            c += 1
    return c

File: experiments/03_synth_data_generation/c_src/batch03_array_ops.c
/* Batch 03: Array operations — uses pointer+length pairs, modeled as Array in Lean */

int sum_array(const int *arr, int n) {
    int s = 0;
    for (int i = 0; i < n; i++) {
        s += arr[i];
    }
    return s;
}

int product_array(const int *arr, int n) {
    int p = 1;
    for (int i = 0; i < n; i++) {
        p *= arr[i];
    }
    return p;
}

int max_element(const int *arr, int n) {
    int m = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > m) m = arr[i];
    }
    return m;
}

int min_element(const int *arr, int n) {
    int m = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] < m) m = arr[i];
    }
    return m;
}

void reverse_array(int *arr, int n) {
    for (int i = 0; i < n / 2; i++) {
        int tmp = arr[i];
        arr[i] = arr[n - 1 - i];
        arr[n - 1 - i] = tmp;
    }
}

int is_sorted(const int *arr, int n) {
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) return 0;
    }
    return 1;
}

int count_occurrences(const int *arr, int n, int val) {
    int c = 0;
    for (int i = 0; i < n; i++) {
        if (arr[i] == val) c++;
    }
    return c;
}
